node/src/components/storage.rs

```diff
@@ -276,7 +276,7 @@ impl Storage {
         Ok(match req {
             StorageRequest::PutBlock { block, responder } => {
                 let mut txn = self.env.begin_rw_txn()?;
-                let outcome = txn.put_value(self.block_db, block.hash(), &block, false)?;
+                let outcome = txn.put_value(self.block_db, block.hash(), &block, true)?;
```

Comment:
@Fraser999 @marc-casperlabs Can you confirm that it is safe to do? We need to be able to overwrite a block since we want to add proofs (finality signatures). I think that the old version allowed for overwrites since we were able to fetch blocks with proofs from the storage and signature has always been added after the block was already stored.

Reply:
I believe you're correct here, although I'd like to have Marc's input on this.  We did discuss this point during the rework, but perhaps it was overlooked or maybe he had planned to address this soon in a follow up whereby we don't need to overwrite whole blocks just to append a signature.

If you don't hear otherwise though, assume it's safe to merge your change.  I can't see a problem other than the efficiency point.

Reply:
OK, thanks.

Reply:
I don't think we should be mutating blocks at all, to be honest. Can we not store finality signature next to blocks? Mid-term, I would like to see the storage be full of immutable objects, everything else is a headache.

We could store all finality signature separate in a loose database and create an index, mapping a block hash to a list of hashes for finality signatures for example.

Reply:
I can create a follow-up PR to do this but we already have a regression in the code where client requesting a block won't see block's proof.

Reply:
> I also don't like that we are reinventing indexes (building them manually) rather than using a database that has them built-in (like SQLite).

Indices aren't exclusive to relational databases though. We also don't need to create queries on the fly and want to take advantage of the memory-mapped performance benefits, so just stuffing SQLite in there is likely not the way to go.

I guess we can postpone this and refactor later.